modules/python/python-http-header.c

```diff
@@ -0,0 +1,399 @@
+  if (!py_ret_list)
+    {
+      gchar buf[256];
+
+      msg_error("Invalid response returned by Python call",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto cleanup;
+    }
+
+  msg_debug("Python call returned valid response",
+            evt_tag_str("class", self->class),
+            evt_tag_str("method", "get_headers"),
+            evt_tag_str("return_type", py_ret_list->ob_type->tp_name));
+
+  if (!_py_append_pylist_to_list(py_ret_list, &headers))
+    {
+      gchar buf[256];
+
+      msg_error("Converting Python List failed",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+      goto cleanup;
+    }
+
+  data->result = HTTP_HEADER_REQUEST_SLOT_SUCCESS;
+
+cleanup:
+  Py_XDECREF(py_args);
+  Py_XDECREF(py_list);
+  Py_XDECREF(py_ret_list);
+
+  PyGILState_Release(gstate);
+
+  if (headers)
+    {
+      g_list_foreach(headers, _append_str_to_list, data->request_headers);
+      g_list_free_full(headers, g_free);
+    }
+}
+
+static gboolean
+_init(PythonHttpHeaderPlugin *self)
+{
+  PyGILState_STATE gstate = PyGILState_Ensure();
+  _py_perform_imports(self->loaders);
+
+  if (!_py_attach_bindings(self))
+    goto fail;
+
+  PyGILState_Release(gstate);
+  return TRUE;
+
+fail:
+  PyGILState_Release(gstate);
+  return FALSE;
+}
+
+static gboolean
+_attach(LogDriverPlugin *s, LogDriver *driver)
+{
+  PythonHttpHeaderPlugin *self = (PythonHttpHeaderPlugin *) s;
+
+  if (!_init(self))
+    {
+      msg_error("Plugin initialization failed",
+                evt_tag_str("plugin", PYTHON_HTTP_HEADER_PLUGIN));
+
+      return FALSE;
+    }
+
+  SignalSlotConnector *ssc = driver->super.signal_slot_connector;
+
+  CONNECT(ssc, signal_http_header_request, _append_headers, self);
+
+  msg_debug("SignalSlotConnector slot registered",
+            evt_tag_printf("signal", "%p", ssc),
+            evt_tag_printf("plugin_name", "%s", PYTHON_HTTP_HEADER_PLUGIN),
+            evt_tag_printf("plugin_instance", "%p", s));
+
+  return TRUE;
+}
+
+static void
+_free(LogDriverPlugin *s)
+{
+  PythonHttpHeaderPlugin *self = (PythonHttpHeaderPlugin *) s;
+
+  g_free(self->class);
+
+  if (self->options)
+    g_hash_table_unref(self->options);
+
+  if (self->loaders)
+    g_list_free_full(self->loaders, g_free);
+
+  PyGILState_STATE gstate = PyGILState_Ensure();
+  _py_detach_bindings(self);
+  PyGILState_Release(gstate);
+
+  log_driver_plugin_free_method(s);
+}
+
+PythonHttpHeaderPlugin *
+python_http_header_new(void)
+{
+  PythonHttpHeaderPlugin *self = g_new0(PythonHttpHeaderPlugin, 1);
+
+  log_driver_plugin_init_instance(&(self->super), PYTHON_HTTP_HEADER_PLUGIN);
+
```

Comment:
I would prefer to have mark-errors-as-critical to be true by default, that should be set somewhere around here.

Reply:
Yes, it could be better as default.